docs/content/docs/connectors/datastream/kafka.md

```diff
@@ -38,10 +38,215 @@ For details on Kafka compatibility, please refer to the official [Kafka document
+    // Start from earliest offset
+    .setStartingOffsets(OffsetsInitializer.earliest())
+    // Start from latest offset
+    .setStartingOffsets(OffsetsInitializer.latest())
+```
+
+You can also implement a custom offsets initializer if built-in initializers above cannot fulfill
+your requirement.
+
+If offsets initializer is not specified, **OffsetsInitializer.earliest()** will be
+used by default.
+
+### Boundedness
+Kafka source is designed to support both streaming and batch running mode. By default, the KafkaSource
+is set to run in streaming manner, thus never stops until Flink job fails or is cancelled. You can use
+```setBounded(OffsetInitializer)``` to specify stopping offsets and set the source running in
+batch mode. When all partitions have reached their stooping offsets, the source will exit.
+
+You can also set KafkaSource running in streaming mode, but still stop at the stopping offset by
+using ```setUnbounded(OffsetInitializer)```. The source will exit when all partitions reach their
+specified stopping offset.
+
+### Additional Properties
+In addition to properties described above, you can set arbitrary properties for KafkaSource and
+KafkaConsumer by using ```setProperties(Properties)``` and ```setProperty(String, String)```.
+KafkaSource has following options for configuration:
+- ```client.id.prefix``` defines the prefix to use for Kafka consumer's client ID
+- ```partition.discovery.interval.ms``` defines the interval im milliseconds for Kafka source
+  to discover new partitions. See <a href="#dynamic-partition-discovery">Dynamic Partition Discovery</a>
+  below for more details.
+
+For configurations of KafkaConsumer, you can refer to
+<a href="http://kafka.apache.org/documentation/#consumerconfigs">Apache Kafka documentation</a>
+for more details.
+
+Please note that the following keys will be overridden by the builder even if
+it is configured:
+- ```key.deserializer``` is always set to ```ByteArrayDeserializer```
+- ```value.deserializer``` is always set to ```ByteArrayDeserializer```
+- ```auto.offset.reset.strategy``` is overridden by ```OffsetsInitializer#getAutoOffsetResetStrategy()```
+  for the starting offsets
+- ```partition.discovery.interval.ms``` is overridden to -1 when
+  ```setBounded(OffsetsInitializer)``` has been invoked
+
+The code snippet below shows configuring KafkaConsumer to use "PLAIN" as SASL mechanism and provide
+JAAS configuration:
+```java
+KafkaSource.builder()
+    .setProperty("sasl.mechanism", "PLAIN")
+    .setProperty("sasl.jaas.config", "org.apache.kafka.common.security.plain.PlainLoginModule required username=\"username\" password=\"password\";")
+```
+
+### Dynamic Partition Discovery
+In order to handle scenarios like topic scaling-out or topic creation without restarting the Flink
+job, Kafka source can be configured to periodically discover new partitions under provided 
+topic-partition subscribing pattern. The discovery is disabled by default. To enable this feature, 
```

Comment:
Can we highlight that "The discovery is disabled by default"? It is important for users to know this.

Reply:
Added a warning box in ce7e539